### PR TITLE
status: catch a few more Progressing cases

### PR DESCRIPTION
**Pull request by squeed**
The easiest way to tell that a daemonset is progressing is that UpdatedNumberScheduled is less than Desired.

Also, catch the case where the daemonset update hasn't yet been observed by the controller.

### CODE REVIEW

**Comment by squeed** [comment]
/hold
Wanted to submit this for review now, but ran out of time to completely test it.

**Comment by squeed** [comment]
context: Clayton noticed in https://bugzilla.redhat.com/show_bug.cgi?id=1695200 that LastTransitionTime for Progressing wasn't bumped after an update.

File: pkg/controller/statusmanager/status_manager_test.go [hunk 1]
```diff
@@ -307,6 +308,16 @@ func TestStatusManagerSetFromDaemonSets(t *testing.T) {
 		t.Fatalf("unexpected Status.Conditions: %#v", co.Status.Conditions)
 	}
 
+	progressingTS := metav1.Now()
+	for _, cond := range co.Status.Conditions {
+		if cond.Type == configv1.OperatorProgressing {
```
Comment (danwinship) [inline pkg/controller/statusmanager/status_manager_test.go]: cond := v1helpers.FindStatusCondition(co.Status.Conditions, configv1.OperatorProgressing)

File: pkg/controller/statusmanager/status_manager.go [hunk 2]
```diff
@@ -195,6 +195,10 @@ func (status *StatusManager) SetFromPods() {
 			progressing = append(progressing, fmt.Sprintf("DaemonSet %q is not available (awaiting %d nodes)", dsName.String(), ds.Status.NumberUnavailable))
 		} else if ds.Status.NumberAvailable == 0 {
 			progressing = append(progressing, fmt.Sprintf("DaemonSet %q is not yet scheduled on any nodes", dsName.String()))
+		} else if ds.Status.UpdatedNumberScheduled < ds.Status.DesiredNumberScheduled {
+			progressing = append(progressing, fmt.Sprintf("DaemonSet %q update is rolling out (%d out of %d updated)", dsName.String(), ds.Status.UpdatedNumberScheduled, ds.Status.DesiredNumberScheduled))
+		} else if ds.Generation > ds.Status.ObservedGeneration {
```
Comment (danwinship) [inline pkg/controller/statusmanager/status_manager.go]: The DaemonSet documentation is egregiously ambiguous, but it seems like it should not be possible to hit this case, given the previous case?

Also, we probably need a corresponding change to the Deployment-tracking code?

Comment (squeed) [reply pkg/controller/statusmanager/status_manager.go]: I played through an upgrade, and saw this case (generation > observedGeneration) after the operator (well, me at the CLI) had updated the daemonset and before the daemonset controller had reacted to it. 

And, yes, the deployment controller also needs some tweaks.

**Comment by squeed** [comment]
Still not sure if / how to catch the (admittedly unlikely) case where we update a DS then go offline until the controller has completely rolled out the change. The only thing I can think of is persisting observedgeneration somewhere, which feels yucky.

**Comment by squeed** [comment]
Updated, tested with a running cluster and verified that progressing had a useful status. Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1695200

/hold cancel
@danwinship PTAL, thanks!

File: pkg/controller/statusmanager/status_manager.go (hunk 2 above)
Comment (danwinship) [reply pkg/controller/statusmanager/status_manager.go]: > after the operator had updated the daemonset and before the daemonset controller had reacted to it.

right, but that's very transient; the daemonset controller should then react to it, at which point `ds.Status.UpdatedNumberScheduled < ds.Status.DesiredNumberScheduled` will be true

**Comment by danwinship** [comment]
> Still not sure if / how to catch the (admittedly unlikely) case where we update a DS then go offline until the controller has completely rolled out the change. The only thing I can think of is persisting observedgeneration somewhere, which feels yucky.

Hm... compare `ds.Status.Conditions[something].LastTransitionTime` with operator.Status.Conditions["Progressing"].LastTransitionTime?

**Comment by squeed** [comment]
> Hm... compare `ds.Status.Conditions[something].LastTransitionTime` with operator.Status.Conditions["Progressing"].LastTransitionTime?

It does't look like the DaemonSet controller is actually setting any `Status.Condition[]` entries yet.

**Pull request merged**